src/index/FTSAlgorithms.cpp

```diff
@@ -306,48 +302,72 @@ void FTSAlgorithms::aggScoresAndTakeTopKContexts(
       << " elements to a table with distinct entities "
       << "and at most " << k << " contexts per entity.\n";
 
+  size_t numOfTerms = dynResult->numColumns() - 3;
+
   // The default case where k == 1 can use a map for a O(n) solution
   if (k == 1) {
-    aggScoresAndTakeTopContext<3>(wep, dynResult);
+    aggScoresAndTakeTopContext<WIDTH>(wep, dynResult);
     return;
   }
 
   // Use a set (ordered) and keep it at size k for the context scores
   // This achieves O(n log k)
   LOG(DEBUG) << "Heap-using case with " << k << " contexts per entity...\n";
 
-  using ScoreToContextAndWord =
-      std::set<std::tuple<Score, TextRecordIndex, WordIndex>>;
-  using ScoreAndStCaW = pair<Score, ScoreToContextAndWord>;
-  using AggMap = ad_utility::HashMap<Id, ScoreAndStCaW>;
-  AggMap map;
+  using AggMapContextAndEntityToWord =
```

Comment:
This is the place where I stopped the review for now (it is quite a lot of code).
I will continue tomorrow. 
I haven't looked at the rest of this file yet and not at the `FTSAlgorithmsTest.cpp`.